witness/internal/witness/witness.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/logverifier"
+	"golang.org/x/mod/sumdb/note"
+)
+
+type Chkpt struct {
+	Parsed	*log.Checkpoint
+	Raw	[]byte
+}
+
+type ChkptStorage interface {
+	// GetLatest returns the latest checkpoint for a given log.
+	GetLatest(logPK string) (*Chkpt, error)
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log.
+	SetCheckpoint(logPK string, c Chkpt) error
+}
+
+type Witness struct {
+	db       ChkptStorage
+	Signer	 note.Signer
+	Hasher   hashers.LogHasher
+	LogV	 logverifier.LogVerifier
+	SigVs	 []note.Verifier
+}
+
+// NewWitness creates a new witness, which initially has no logs to follow.
+func NewWitness(h hashers.LogHasher, db ChkptStorage, sk string) (*Witness, error) {
+	ns, err := note.NewSigner(sk)
+	if err != nil {
+		return nil, err
+	}
+	return &Witness{
+		db:       db,
+		Signer:	  ns,
+		Hasher:   h,
+		LogV:	  logverifier.New(h),
+		SigVs:    []note.Verifier{},
+	}, nil
+}
+
+// parse verifies the checkpoint under the given list of possible public keys.  
+// It returns the parsed checkpoint and the list of public keys under which the 
+// signature(s) verified.
+func (w *Witness) parse(chkptRaw []byte, verifiers []note.Verifier) (*log.Checkpoint, []string, error) {
+	n, err := note.Open(chkptRaw, note.VerifierList(verifiers...))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	chkpt := &log.Checkpoint{}
+	_, err = chkpt.Unmarshal([]byte(n.Text))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal new checkpoint: %w", err)
+	}
+	var keys []string
+	for _, sig := range n.Sigs {
+	    keys = append(keys, sig.Name)
+	}
+	return chkpt, keys, nil
+}
+
+// GetLatest gets the latest checkpoint for a given log, which should be
+// consistent with all other checkpoints for the same log.  It also signs it
+// under the witness' key.
+func (w *Witness) GetLatest(logPK string) ([]byte, error) {
+	chkpt, err := w.db.GetLatest(logPK)
+	if err != nil {
+		return nil, err
+	}
+	// Add the witness' signature to the checkpoint.
+	// TODO find a way so we don't have to verify the signatures again.
+	n, err := note.Open(chkpt.Raw, note.VerifierList(w.SigVs...))
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	cosigned, err := note.Sign(n, w.Signer)
+	if err != nil {
+		return nil, err
+	}
+	return cosigned, nil
+}
+
+// Update updates latest checkpoint if `to` matches the current version (`from`).
+func (w *Witness) Update(fromRaw, toRaw []byte, proof log.Proof) error {
+	// Check the signatures on both of the raw checkpoints and parse them
+	// into the log.Checkpoint format.
+	from, keysFrom, err := w.parse(fromRaw, w.SigVs)
+	if err != nil {
+		return err
+	}
+	to, keysTo, err := w.parse(toRaw, w.SigVs)
+	if err != nil {
+		return err
+	}
+	// Make sure both verified with respect to the same public key.
+	if len(keysFrom) > 1 || len(keysTo) > 1 {
```

Comment:
Should we handle the case where `to` was already cosigned by another witness, or do we expect the log to send us "pristine" checkpoints and do the cosign merge on their end? (maybe this is a TODO!)

Reply:
Yeah, this is mostly assuming the latter but I can add more flexible arrangements as a TODO.  I guess even in the witness co-signed case it should still be the case that the list is of size 1 since witnesses should have only the public keys of logs (and not of other witnesses), meaning the signatures from other witnesses would go unverified?

Reply:
Ah yes, a log and witness sharing a key would lead to much "excitement" :)